### PR TITLE
Refactor the model for the 2022 elections

### PR DESCRIPTION
**Pull request by rlouf**
In this PR I refactor the code for the 2022 elections model with two goals in mind: make it easier to modify, and easier to read for someone external to the projects.

This is work in progress, and progress will be made so that the model can still run at every commit.

## TODO

- [x] Document the model inline
- [ ] Write mathematical model for #pollspositions/presidentielles
- [ ] Give more explicit names to variables whenever needed
- [ ] Add function to save traces
- [ ] Factor the data cleaning process
- [ ] Can we remove some of these notebooks?

### CODE REVIEW

File: presidential-elections/utils/model.py [hunk 1]
```diff
@@ -423,19 +436,23 @@ def build_model(
...
-        polls: pd.DataFrame, default None
-            All raw polls from past and current elections.
-            Specify only if you want to run the model on another dataset than the training one (e.g
-            for out-of-sample predictions).
-        continuous_predictors: pd.DataFrame, default None
-            A dataframe of all the continuous predictors.
-            Specify only if you want to run the model on another dataset than the training one (e.g
-            for out-of-sample predictions).
+        polls
+            Poll results from past and current elections. This only need to be
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:444]:
```suggestion
            Poll results from past and current elections. This only needs to be
```

File: presidential-elections/utils/model.py [hunk 2]
```diff
@@ -423,19 +436,23 @@ def build_model(
...
+            specified for out-of-sample predictions to run the model on another
+            dataset than the training data.
+        continuous_predictors
+            Continuous predictors, or fundamentals. This only need to be
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:448]:
```suggestion
            Continuous predictors, or fundamentals. This only needs to be
```

File: presidential-elections/utils/model.py [hunk 3]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
+            # --------------------------------------------------------
+
+            # Baseline latent popularity for each political family. Shared
+            # across elections.
+            party_intercept_sd = pm.HalfNormal("party_intercept_sd", 0.5)
+            party_intercept = ZeroSumNormal(
+                "party_intercept", sigma=party_intercept_sd, dims="parties_complete"
+            )
+
+            # Election-specific latent popularity for each political family.
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:481]:
```suggestion
            # Election-specific deviation from baseline latent popularity for each political family.
```

File: presidential-elections/utils/model.py [hunk 4]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
+                    zerosum_axes=(0, 1),
+                )
+            )
+
+            # --------------------------------------------------------
+            #                        HOUSE EFFECTS
+            # --------------------------------------------------------
+
+            # Baseline for polls' bias towards the different political families.
+            # These biases are shared by pollsters.
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:509]:
```suggestion
            # These biases are shared by pollsters, i.e bias of the market
```

File: presidential-elections/utils/model.py [hunk 5]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
+            # --------------------------------------------------------
+            #                  FUNDAMENTAL COMPONENT
+            #
+            # It is commonly assumed that the results of the elections
+            # are mostly determined by economic fundamentals, and
+            # that the opinion "drifts" towards this result during the
+            # campaign, so to speak.
+            #
+            # The coefficient below accounts for the effect of the
+            # unemployment in the election result.
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:552]:
```suggestion
            # The coefficient below accounts for the effect of
            # unemployment on the election result.
```

File: presidential-elections/utils/model.py [hunk 6]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
+            # the course of an election. We model this evolution with
+            # gaussian processes.
+            #
+            # We currently use gaussian processes with 3 different
+            # lengthscales to account for the typical timescales over which
+            # opinion can change.
+            #
+            # The time evolution has two components: one that depends only
+            # on the current elections and another than is common to all
+            # elections (the baseline).
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:574]:
```suggestion
            # The time evolution has two components: one that is common to all
            # elections (the baseline), and another one for each election, which is
            # a deviation from the common baseline.
```

File: presidential-elections/utils/model.py [hunk 7]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
             )
+
+            # --------------------------------------------------------
+            #                    ELECTION RESULTS
+            #
+            # In this section we use the variables defined before to model the
+            # political families' latent popularity and how it translates into
+            # results the day of the election.
+            #
+            # Fundamental data only enter the model here; poll results enter
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:728]: > Fundamental data only enter the model here;

Mmmh no, fundamentals are already used in `latent_mu` above. This is exactly the same regression, except only at time 0. It's just that that timestep is informed by election results, whereas previous timesteps are informed by polls

File: presidential-elections/utils/model.py [hunk 8]
```diff
@@ -449,50 +466,299 @@ def build_model(
...
+            # The mean (1000) is thus taken to be roughly the sample size of
+            # polls, and the standard deviation accounts for the variation in
+            # sample size.
+            concentration_results = pm.InverseGamma(
+                "concentration_results", mu=1000, sigma=200
+            )
             pm.DirichletMultinomial(
                 "R",
-                a=concentration * latent_pop_t0[:-1],
+                a=concentration_results * latent_pop_t0[:-1],
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:761]: I understand what you're doing here, and I tried that already but it really didn't work, as we don't have enough past election to really learn `concentration_results`, so the model's convergence and predictions were really bad. Did you run the model with this parametrization? Did it work?

In any case, I don't think it's a huge deal to use the same concentration for polls _and_ results, as they describe the same underlying process (and we're not using the _real_ sample size for election results anyways)

File: presidential-elections/utils/model.py [hunk 9]
```diff
@@ -581,239 +847,15 @@ def _build_data_containers(
...
-            ),
-        )
-
     def sample_all(
         self, *, model: pm.Model = None, var_names: List[str], **sampler_kwargs
     ) -> arviz.InferenceData:
         """
         Sample the model and return the trace.
+
+        TODO: Add 3 distinct functions to sample from the prior predictive,
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:856]: I think you can just add kwargs here

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
Thanks @rlouf ! Left my comments. 
One thing: why are you removing all the intermediate functions that build the model? I like those because they give you a quick idea of the model's overall structure, and then you can dive into each function -- which is even easier now with your comments

**Comment by AlexAndorra** [comment]
> why does the regression that is conditioned on the poll results takes the unemployment as a covariate? 

The thinking is: polls and election results are two similar ways of approaching the latent popularity of each party (that we never observe in reality). At the same time, unemployment is assumed to be correlated with those three elements -- polls, election results and true latent popularity. So, learning the correlation patterns with the first two elements helps us learn the pattern for the last one, which is the one we care about.

In short: election results have nothing special, they are just a much bigger poll with no bias (which gives us more info about the true latent popularity).

Is that clearer?

**Comment by rlouf** [comment]
> Thanks @rlouf ! Left my comments. 

Thanks

> One thing: why are you removing all the intermediate functions that build the model?

Because imho they made the code harder to understand. You'll notice I rearranged the code a little to help understand what's going on and this would not have been possible with the current functions.

**Comment by rlouf** [comment]
> Is that clearer?

Yes that makes sense. Will add that answer to the mathematical description of the model in pollsposition/presidentielles.

**Comment by rlouf** [comment]
Can we delete some on the notebooks btw?

File: presidential-elections/utils/model.py (hunk 7 above)
Comment (rlouf) [reply presidential-elections/utils/model.py:728]: True, I meant election results. Good catch.

File: presidential-elections/utils/model.py (hunk 8 above)
Comment (rlouf) [reply presidential-elections/utils/model.py:761]: This will need further investigation before merging, but I don't have any convergence issue with separate concentration parameters. I will have a look at these parameters' posterior distribution.

File: presidential-elections/utils/model.py (hunk 9 above)
Comment (rlouf) [reply presidential-elections/utils/model.py:856]: Rémi not like kwargs when he can avoid them. We can keep this function, however.

File: presidential-elections/utils/model.py [hunk 10]
```diff
@@ -39,45 +39,50 @@ class PresidentialElectionsModel:
...
     """
 
+    political_families = [
+        "farleft",
+        "left",
+        "green",
+        "center",
+        "right",
+        "farright",
+        "souverainistes",
```
Comment (AlexAndorra) [inline presidential-elections/utils/model.py:49]: Are you down to rename "souverainistes" to "identitaires" in this PR?

**Comment by AlexAndorra** [comment]
> Can we delete some on the notebooks btw?

You can just keep `dynamic-model-2022` and `dynamic-model-final` in `notebooks`, and put all the rest in a new sub-directory, e.g `old`

File: presidential-elections/utils/model.py (hunk 8 above)
Comment (rlouf) [reply presidential-elections/utils/model.py:769]: They do!

Comment (AlexAndorra) [reply presidential-elections/utils/model.py:769]: Nice, then we can add it! Feels nice when we solve something without realizing it 😆

File: presidential-elections/utils/model.py (hunk 10 above)
Comment (rlouf) [reply presidential-elections/utils/model.py:49]: It looks like `souverainistes` is used in the file with the election results as well. Should be open a separate issue?

Comment (AlexAndorra) [reply presidential-elections/utils/model.py:60]: Yes please

**Comment by rlouf** [comment]
Only refactored the model, merging before the PR gets out of hand.
